Add UDP GSO/GRO support (Linux) and --gsro switch

gegles opened the pull request:
UDP GSO/GRO Support - Refactored Implementation

This PR implements UDP Generic Segmentation Offload (GSO) and Generic Receive Offload (GRO) support for iperf3 on Linux, addressing all maintainer feedback with a cleaned-up implementation.

## Changes

### 1. Feature Enablement (Disabled by Default)
- Changed from --no-gsro to --gsro flag
- GSO/GRO is now opt-in rather than opt-out
- Default behavior unchanged for existing users

### 2. Code Refactoring
- Eliminated ~84 lines of duplicated code
- Unified packet processing loops in iperf_udp_send() and iperf_udp_recv()
- Single code path handles both GSO/GRO-enabled and disabled cases
- Improved maintainability and reduced error potential

### 3. Bug Fix: Missing Counters
- Critical fix: Loss and jitter counters now work in ALL modes
- Previously only calculated when GRO was enabled
- Moved counter calculations outside #ifdef guards
- Users without --gsro now get proper statistics

### 4. Debug Output Control
- Gated GSO/GRO diagnostic messages behind test->debug flag
- Cleaner output for production use
- Debug info still available with -d flag

### 5. Documentation
- Added --gsro flag to iperf3.1 man page
- Documented GSO/GRO functionality and platform support

## Testing

On 2 bare-metal Linux machines connected with 100Gbps NICs I get:

Normal mode (no GSO/GRO):
```
./src/iperf3 -c rapid10-100g -u -b 35G
Connecting to host rapid10-100g, port 5201
[  5] local 192.168.100.9 port 47526 connected to 192.168.100.10 port 5201
[ ID] Interval           Transfer     Bitrate         Total Datagrams
[  5]   0.00-1.00   sec   296 MBytes  2.48 Gbits/sec  214425  
[  5]   1.00-2.00   sec   294 MBytes  2.47 Gbits/sec  212846  
[  5]   2.00-3.00   sec   291 MBytes  2.44 Gbits/sec  210867  
[  5]   3.00-4.00   sec   291 MBytes  2.44 Gbits/sec  210949  
[  5]   4.00-5.00   sec   292 MBytes  2.45 Gbits/sec  211456  
[  5]   5.00-6.00   sec   296 MBytes  2.48 Gbits/sec  214033  
[  5]   6.00-7.00   sec   296 MBytes  2.49 Gbits/sec  214711  
[  5]   7.00-8.00   sec   296 MBytes  2.48 Gbits/sec  214328  
[  5]   8.00-9.00   sec   296 MBytes  2.48 Gbits/sec  214206  
[  5]   9.00-10.00  sec   296 MBytes  2.48 Gbits/sec  214480  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-10.00  sec  2.88 GBytes  2.47 Gbits/sec  0.000 ms  0/2132301 (0%)  sender
[  5]   0.00-10.00  sec  2.88 GBytes  2.47 Gbits/sec  0.002 ms  0/2132301 (0%)  receiver
```
So `2.47 Gbits/sec` max (even though 35G requested)

Now with `--gsro` enabled:

```
./src/iperf3 -c rapid10-100g -u -b 35G --gsro
Connecting to host rapid10-100g, port 5201
[  5] local 192.168.100.9 port 56515 connected to 192.168.100.10 port 5201
[ ID] Interval           Transfer     Bitrate         Total Datagrams
[  5]   0.00-1.00   sec  4.08 GBytes  35.0 Gbits/sec  3024450  
[  5]   1.00-2.00   sec  4.07 GBytes  35.0 Gbits/sec  3021435  
[  5]   2.00-3.00   sec  4.07 GBytes  35.0 Gbits/sec  3021255  
[  5]   3.00-4.00   sec  4.07 GBytes  35.0 Gbits/sec  3021434  
[  5]   4.00-5.00   sec  4.07 GBytes  35.0 Gbits/sec  3021526  
[  5]   5.00-6.00   sec  4.07 GBytes  35.0 Gbits/sec  3021390  
[  5]   6.00-7.00   sec  4.07 GBytes  35.0 Gbits/sec  3021435  
[  5]   7.00-8.00   sec  4.07 GBytes  35.0 Gbits/sec  3021390  
[  5]   8.00-9.00   sec  4.07 GBytes  35.0 Gbits/sec  3019132  
[  5]   9.00-10.00  sec  4.08 GBytes  35.0 Gbits/sec  3023603  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-10.00  sec  40.7 GBytes  35.0 Gbits/sec  0.000 ms  0/30217050 (0%)  sender
[  5]   0.00-10.00  sec  40.7 GBytes  35.0 Gbits/sec  0.000 ms  0/30217005 (0%)  receiver
```

Result: `35.0 Gbits/sec` max throughput achieved. Beyond that some packet loss starts to happen due to CPU bound. On more modern machines more could be achieved.

## Benefits
- Reduced CPU overhead for high-throughput UDP tests
- Improved throughput on supported hardware
- Cleaner, more maintainable codebase
- Proper statistics in all modes

marcosfsch committed:
add UDP GSO/GRO support on Linux and --no-gsro option

This change adds first-class support for Linux UDP Generic Segmentation
Offload (GSO) and Generic Receive Offload (GRO) in iperf3.

At configure time, the build detects availability of the UDP_SEGMENT
and UDP_GRO socket options via <linux/udp.h> and enables code paths
accordingly. On capable systems, these features are now enabled by
default for UDP tests.

A new command-line flag, --no-gsro, allows users to disable GSO and GRO
even when supported by the kernel. Help text is included in usage_longstr.

Additional changes:
 - Updated iperf_settings to track GSO/GRO state and buffer/segment sizes.
 - Added a warning if the configured UDP block size exceeds the TCP MSS.
 - Ensured behavior is unchanged on systems without GSO/GRO support.

GSO can reduce CPU overhead on send by offloading UDP segmentation to
the kernel/NIC. GRO can reduce per-packet processing cost on receive by
coalescing incoming UDP segments. Together they can improve throughput
and efficiency in high-rate UDP tests on modern Linux systems.

# Conflicts:
#	src/iperf_api.c
#	src/iperf_api.h

gegles commented:
FYI @bmah888 or @jefposkanzer, any way someone can review and hopefully merge this?

gegles committed:
fix(GRO): fix per-datagram parsing and loss accounting

Parse coalesced GRO payloads using the negotiated blksize stride and account loss/jitter per datagram. Avoids inflated “loss” when kernel GRO hints are unreliable. No change to GSO send behavior.

davidBar-On commented on src/iperf_api.c:
```diff
@@ -1835,6 +1851,20 @@ iperf_parse_arguments(struct iperf_test *test, int argc, char **argv)
+            test->settings->gso_bf_size = (test->settings->gso_bf_size / test->settings->gso_dg_size) * test->settings->gso_dg_size;
+        } else {
+            /* If gso_dg_size is 0 (unlimited bandwidth), use default UDP datagram size */
+            test->settings->gso_dg_size = 1472; /* Standard UDP payload size for Ethernet MTU */
```
Why not using `DEFAULT_UDP_BLKSIZE` instead of 1472?  If it is important that 1472 will be used, suggest to add `#define DEFAULT_GSO_DG_BLKSIZE 1472`.

gegles commented on src/iperf_api.c:
```diff
@@ -1714,6 +1717,17 @@ iperf_parse_arguments(struct iperf_test *test, int argc, char **argv)
 		break;
+#endif
+#if defined(HAVE_UDP_SEGMENT) || defined(HAVE_UDP_GRO)
+            case OPT_NO_GSRO:
```
@davidBar-On I hear what you’re saying, and I’m open to changing the option’s behavior. For example, I originally had two distinct options to explicitly enable or disable each feature (`--no-gso` and `--no-gro`). But then I started thinking: if the kernel supports them and they’re implemented correctly, why wouldn’t we want both enabled by default? The packets on the wire are identical, and the benefits—whether higher throughput or lower CPU usage—are significant.

That said, if the consensus is to provide more knobs and finer-grained control, I’m fine with that too.

gegles committed:
udp gso/gro: replace hard-coded 1472 with DEFAULT_UDP_BLKSIZE fallback

Use the existing DEFAULT_UDP_BLKSIZE for GSO datagram-size fallback instead of the literal 1472.

Rationale: avoids a magic number and keeps a conservative, widely safe default across IPv4/IPv6 when the control socket MSS is unavailable. In normal operation UDP blksize is derived from the control TCP MSS; this constant fallback is only used when MSS cannot be determined or when the computed gso_dg_size ends up 0 (unlimited case).

Behavior notes:

- If the user sets -l/--length, that value drives both UDP block size and gso_dg_size.

- Otherwise, gso_dg_size tracks the chosen blksize; it falls back to DEFAULT_UDP_BLKSIZE only when the computed value is 0.

Files:

- src/iperf_api.c: update two fallback sites

- src/iperf_client_api.c: update fallback site

gegles committed:
udp gso/gro: move policy to client; send explicit params; server accepts and applies locally

Client remains the source of truth for UDP GSO/GRO policy (including --no-gsro). During parameter exchange, the client now sends GSO/GRO flags and sizes in JSON, and the server simply consumes those values without recomputing.

Kernel capability gating stays local and authoritative: each endpoint attempts to enable GSO/GRO on its own sockets via setsockopt, and if the kernel rejects it, we log and flip the local flag off. This handles the case where only one side supports the feature (GSO for the sender, GRO for the receiver).

Backward compatibility: if talking to an older client that doesn’t send GSO fields, the server derives gso_dg_size from blksize and adjusts gso_bf_size, falling back to DEFAULT_UDP_BLKSIZE if zero. This preserves previous behavior without overriding explicit client intent.

Behavior details:

- --no-gsro on the client sends gso=0 and gro=0, so the server won’t try to enable them.

- If -l/--length is provided, blksize (and therefore gso_dg_size when enabled) follows that value; otherwise default logic applies.

Files: src/iperf_api.c (send_parameters adds gso/gro fields; get_parameters reads them and removes server-side recompute unless needed for compatibility).

gegles commented:
@swlars, just FYI, I rebased on latest main (no changes, just a minor conflict resolution.

Any idea when you'll be able to review/merge this?

LMK. thx!

bmah888 commented:
Hi @gegels, thanks for the additional explanations!

After some more discussion we've decided this yes this PR is something
that we'd like to include in iperf3. Thanks for the education and of
course the PR itself.

There are a few things that we need to address. We can help with some
of these.

1. GSO/GRO functionality really should be off by default. I think it's
   fine to have a single flag (`--gsro`?).
   
2. Manual page should be added.

3. Code duplication in `iperf_udp_recv()` and `iperf_udp_send()`.

4. Missing counter increments for errors and jitter computation. 

I have an idea for solving 3 and 4. We keep the loop for filling (and
checking) the iperf3 payloads from a buffer, but we set the various
loop limits so that if we aren't doing GSO or GRO we go through the
loop exactly one time corresponding to the non-GSO, non-GRO behavior.

Basically we do something like this (this is for `iperf_udp_send()`).

```
size_t dgram_sz;
size_t buf_sz;
char *degram_buf;
char *dgram_buf_end;
#ifdef HAVE_UDP_SEGMENT 
const int min_pkt_size = ...
if (sp->test->settings->gso) {
    /* GSO enabled, pack a buffer with multiple iperf3 payloads */
}
else {
    /* GSO available but disabled, make the buffer one payload size */
    dgram_sz = buf_sz = size;
}
#else
    /* GSO not available, make the buffer one payload size */
    dgram_sz = buf_sz = size;
#endif

dgram_buf = sp->buffer;
dgram_buf_end = dgram_buf + dgram_sz;

/* all the loop parameters should be set accordingly by this point */
while (buf_sz > 0 && dgram_buf + dgram_sz <= dgram_buf_end) {
    /* all the stuff that was in the loop */
}
```

The idea for `iperf_udp_recv()` would be similar...do the receive
operation and setup parameters accordingly. Again we keep the
functionality in the loop (which is moved outside of of conditional
compilation), which includes updating loss counters and jitter
computation.

5. The output in `iperf_udp_gso()` and `iperf_udp_gro()` should be
   gated on `test->debug` or go away.

gegles committed:
udp gso/gro: change to disabled-by-default with --gsro flag

Change from enabled-by-default with --no-gsro to disabled-by-default
with --gsro flag. This makes GSO/GRO opt-in rather than opt-out.

Changes:
- Rename --no-gsro to --gsro flag
- Change default initialization from enabled (1) to disabled (0)
- Update option handler to enable instead of disable
- Update help text to reflect new behavior

gegles committed:
docs: add --gsro flag to iperf3.1 man page

Add documentation for the --gsro flag in the manual page,
describing UDP GSO/GRO functionality and its benefits.

gegles committed:
udp gso/gro: refactor to eliminate duplication and restore counters

Refactor iperf_udp_send() and iperf_udp_recv() to use unified loops
that handle both GSO/GRO-enabled and disabled cases in a single code
path, eliminating code duplication.

Key changes:
- Configure loop parameters (dgram_sz, buf_sz) upfront based on
  GSO/GRO availability
- Use single unified loop for packet processing regardless of
  GSO/GRO state
- Restore loss counter increments and jitter computation outside
  #ifdef guards so they work when GSO/GRO is disabled
- Gate diagnostic output in iperf_udp_gso() and iperf_udp_gro()
  behind test->debug checks to prevent unwanted verbosity

This addresses maintainer feedback to eliminate separate code branches
and ensure counters work correctly in all configurations.

Tested:
- Normal UDP mode: jitter and loss counters working correctly
- --gsro mode: jitter and loss counters working correctly
- Debug output properly gated behind -d flag

gegles committed:
udp gso/gro: make --gsro a client-only option

Reject --gsro when used with -s (server mode) by returning
IECLIENTONLY, matching the pattern used by other client-only flags.
The server already receives GSO/GRO settings from the client via the
JSON parameter exchange, so passing --gsro on the server command line
has no effect. Update help text and man page accordingly.

gegles committed:
udp gso/gro: make --gsro available regardless of local support

Address PR feedback to allow clients without GSO/GRO support to request
server-side enablement. Key changes:

- Remove conditional compilation guards around --gsro option
- Always define gso/gro fields in iperf_settings structure
- Always send/receive GSO/GRO parameters in JSON protocol
- Add warnings when --gsro requested but not supported locally
- Socket options only applied when HAVE_UDP_SEGMENT/HAVE_UDP_GRO defined

This allows a client compiled without GSO/GRO support to still use --gsro
to enable these features on a capable server, improving flexibility for
heterogeneous deployments.

gegles committed:
udp gso/gro: remove remaining conditional guards and restore zerocopy

Remove HAVE_UDP_SEGMENT/HAVE_UDP_GRO guards around:
- GSO parameter initialization in iperf_parse_arguments
- Buffer sizing logic in iperf_new_stream

These fields are now always defined, allowing client/server negotiation
regardless of local support. Guards remain only around actual socket
operations.

Also restore MAP_SHARED for mmap (was accidentally changed to MAP_PRIVATE
during GSO/GRO merge in b56475e), fixing zerocopy functionality that was
broken since PR #1949.

gegles committed:
udp gso/gro: clean up help text guards and debug output

- Remove conditional guards around --gsro help text to match option
  availability (option is always available regardless of local support)
- Remove debug printf statements from net.c (net layer doesn't log to
  console; returns error codes silently per existing pattern)
- Remove duplicate iperf.h include in net.c

This ensures --gsro appears in help on all systems and eliminates
console spam from GSO/GRO error paths.

gegles committed:
udp gso/gro: refactor to use stub implementations and fix issues

Replace compile-time feature guards in headers with stub implementations
that return errors when features are unavailable:

- Remove #ifdef guards from net.h function declarations
- Reorganize net.h to group Nread_gro with read functions and
  Nwrite_gso with write functions
- Add stub implementations in net.c and iperf_udp.c that return errors
  and set gso/gro flags to 0 when features not supported
- Remove guards around ALL function calls (including in iperf_udp_connect)
  to ensure stubs run on platforms without GSO/GRO support
- Keep guards only around setsockopt calls using platform constants

Also fix:
- Unused variable warning: remove cnt from iperf_udp_recv
- Warning ordering: show platform support warnings only after
  client-only check passes (prevents confusing output with -s --gsro)

This fixes the critical bug where --gsro on macOS caused zero traffic
because ifdef guards prevented stubs from running, leaving gso=1 and
triggering the unsupported GSO path.

Benefits:
- Cleaner API (no preprocessor clutter in headers)
- Runtime feature detection via stubs
- Code works correctly on all platforms
- Better code organization and error messages

bmah888 commented:
Thanks for all the revisions! I agree this code is a lot cleaner (and the diff is smaller).

I had a chance to do some testing with the latest version of the PR. No changes to suggest at this point.

Most testing was done between two very old bare-metal x86_64 hosts running Ubuntu 22.04.5 LTS, with Mellanox Connect-X 6 cards connected at 100GE through an L3VPN on a Nokia 7750 SR2S:

* `--gsro` with both IPv4 and IPv6 working as expected
* The combination of `--gsro --reverse` works as expected
* Multiple parallel streams (tested through `--parallel 4`) works as expected
* Slightly less packet loss through 10Gbps with --gsro
* Reproduced the result that at very high -b values (such as -b30g), using --gsro allows the sender to send faster, so the receiver gets more packets and greater aggregate throughput. But this comes at the cost of a higher loss rate.
* Increasing the buffer size (i.e. `-w1m`) can reduce the packet loss rate. This is generally true regardless of whether GSO/GRO are used or not.
* Manual page looks good, thanks for adding that.

On some separate VMs:
* Testing between FreeBSD 14.3 (no GSO/GRO) and Ubuntu 22.04.5 LTS (GSO/GRO enabled) shows they interoperate correctly in both directions.

gegles commented:
> Thanks for all the revisions! I agree this code is a lot cleaner (and the diff is smaller).
> 
> I had a chance to do some testing with the latest version of the PR. No changes to suggest at this point.
> 
> Most testing was done between two very old bare-metal x86_64 hosts running Ubuntu 22.04.5 LTS, with Mellanox Connect-X 6 cards connected at 100GE through an L3VPN on a Nokia 7750 SR2S:
> 
> * `--gsro` with both IPv4 and IPv6 working as expected
> * The combination of `--gsro --reverse` works as expected
> * Multiple parallel streams (tested through `--parallel 4`) works as expected
> * Slightly less packet loss through 10Gbps with --gsro
> * Reproduced the result that at very high -b values (such as -b30g), using --gsro allows the sender to send faster, so the receiver gets more packets and greater aggregate throughput. But this comes at the cost of a higher loss rate.
> * Increasing the buffer size (i.e. `-w1m`) can reduce the packet loss rate. This is generally true regardless of whether GSO/GRO are used or not.
> * Manual page looks good, thanks for adding that.
> 
> On some separate VMs:
> 
> * Testing between FreeBSD 14.3 (no GSO/GRO) and Ubuntu 22.04.5 LTS (GSO/GRO enabled) shows they interoperate correctly in both directions.

This is great news and good feedback... 

But do you confirm that you see massive throughput improvements when GSO/GRO is available (on both sides)?

As I showed above, I go from 2Gbps max to 35Gbps with no packet loss ..... and that's on pretty old machines as well... 

Indeed, if one tries to go above that, then often the sender can overwhelm the receiver and thus generate packet loss...

gegles commented:
Thanks @bmah888, @swlars and @davidBar-On!!!

Very excited that will be officially in `iperf3` and for others to start playing with it!

abajk commented:
@gegles Iperf 3.21 cannot establish a connection with very short packets. Is this a known limitation?
~~~
abajkowski@abajkowski:~$ iperf3 -c 192.168.1.1 -R -t 3600 -u -b 3G -l 500 --gsro
Connecting to host 192.168.1.1, port 5201
Reverse mode, remote host 192.168.1.1 is sending
[  5] local 192.168.1.10 port 55486 connected to 192.168.1.1 port 5201
[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-1.00   sec  0.00 Bytes  0.00 bits/sec  0.000 ms  0/0 (0%)  
^C[  5]   1.00-1.77   sec  0.00 Bytes  0.00 bits/sec  0.000 ms  0/0 (0%)  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Jitter    Lost/Total Datagrams
[  5]   0.00-1.77   sec  0.00 Bytes  0.00 bits/sec  0.000 ms  0/0 (0%)  sender
[  5]   0.00-1.77   sec  0.00 Bytes  0.00 bits/sec  0.000 ms  0/0 (0%)  receiver
iperf3: interrupt - the client has terminated by signal Interrupt(2)
~~~
